File: src/main/java/org/myfintech/payment/exception/handler/ExceptionHandlerConstants.java
package org.myfintech.payment.exception.handler;

/**
 * Constants used in exception handling for consistent error responses.
 * Centralizes all string literals to avoid duplication and improve maintainability.
 */
public final class ExceptionHandlerConstants {
    
    private ExceptionHandlerConstants() {
        // Prevent instantiation
    }
    
    // ================== Property Keys ================== //
    
    /**
     * Standard error response property keys
     */
    public static final String ERROR_ID_KEY = "errorId";
    public static final String TIMESTAMP_KEY = "timestamp";
    public static final String PATH_KEY = "path";
    public static final String ERROR_CODE_KEY = "errorCode";
    public static final String METHOD_KEY = "method";
    public static final String DEBUG_INFO_KEY = "debugInfo";
    
    /**
     * Validation-specific property keys
     */
    public static final String VALIDATION_ERRORS_KEY = "validationErrors";
    public static final String FIELD_ERRORS_KEY = "fieldErrors";
    public static final String GLOBAL_ERRORS_KEY = "globalErrors";
    public static final String FIELD_KEY = "field";
    public static final String MESSAGE_KEY = "message";
    public static final String REJECTED_VALUE_KEY = "rejectedValue";
    public static final String INVALID_VALUE_KEY = "invalidValue";
    public static final String OBJECT_KEY = "object";
    
    /**
     * Type mismatch property keys
     */
    public static final String PARAMETER_KEY = "parameter";
    public static final String EXPECTED_TYPE_KEY = "expectedType";
    public static final String ACTUAL_VALUE_KEY = "actualValue";
    
    /**
     * File processing property keys
     */
    public static final String FILE_NAME_KEY = "fileName";
    
    /**
     * Suggestion property key
     */
    public static final String SUGGESTION_KEY = "suggestion";
    
    // ================== Error Titles ================== //
    
    public static final String RESOURCE_NOT_FOUND_TITLE = "Resource Not Found";
    public static final String BAD_REQUEST_TITLE = "Bad Request";
    public static final String INVALID_ARGUMENT_TITLE = "Invalid Argument";
    public static final String TYPE_MISMATCH_TITLE = "Type Mismatch";
    public static final String VALIDATION_FAILED_TITLE = "Validation Failed";
    public static final String FILE_PROCESSING_ERROR_TITLE = "File Processing Error";
    public static final String FILE_PARSING_ERROR_TITLE = "File Parsing Error";
    public static final String DATA_CONSTRAINT_VIOLATION_TITLE = "Data Constraint Violation";
    public static final String CONCURRENT_UPDATE_CONFLICT_TITLE = "Concurrent Update Conflict";
    public static final String TRANSACTION_FAILED_TITLE = "Transaction Failed";
    public static final String INTERNAL_SERVER_ERROR_TITLE = "Internal Server Error";
    
    // ================== Error Messages ================== //
    
    /**
     * Generic error messages
     */
    public static final String UNEXPECTED_ERROR_MESSAGE = "An unexpected error occurred";
    public static final String UNEXPECTED_ERROR_WITH_NOTIFICATION = "An unexpected error occurred. Our team has been notified.";
    public static final String INVALID_VALUE_DEFAULT = "Invalid value";
    public static final String INVALID_OBJECT_DEFAULT = "Invalid object";
    public static final String UNKNOWN_TYPE = "unknown";
    public static final String UNKNOWN_ENTITY = "Unknown entity";
    public static final String UNKNOWN_LOCATION = "unknown location";
    public static final String UNKNOWN_FILE = "unknown";
    public static final String NULL_VALUE = "null";
    
    /**
     * Validation error messages
     */
    public static final String VALIDATION_CONSTRAINTS_VIOLATED = "One or more validation constraints were violated";
    public static final String REQUEST_VALIDATION_FAILED = "Request validation failed. Please check the errors and try again.";
    public static final String CLIENT_NOT_FOUND = "client not found: ";
    public static final String CONTRACT_NOT_FOUND = "contract not found: ";
    
    /**
     * Data integrity messages
     */
    public static final String DATA_CONSTRAINTS_ERROR = "Unable to process request due to data constraints";
    public static final String DUPLICATE_RECORD_ERROR = "A record with the same information already exists";
    public static final String FOREIGN_KEY_ERROR = "Referenced data not found or cannot be deleted due to existing dependencies";
    
    /**
     * Concurrency messages
     */
    public static final String OPTIMISTIC_LOCK_MESSAGE = "The record was modified by another user. Please refresh and try again.";
    public static final String OPTIMISTIC_LOCK_SUGGESTION = "Refresh the data and retry your changes";
    
    /**
     * Transaction messages
     */
    public static final String TRANSACTION_COMPLETION_ERROR = "Unable to complete the transaction. Please try again.";
    
    // ================== Error Codes ================== //
    
    public static final String RESOURCE_NOT_FOUND_CODE = "RESOURCE_NOT_FOUND";
    public static final String BAD_REQUEST_CODE = "BAD_REQUEST";
    public static final String INVALID_ARGUMENT_CODE = "INVALID_ARGUMENT";
    public static final String TYPE_MISMATCH_CODE = "TYPE_MISMATCH";
    public static final String VALIDATION_FAILED_CODE = "VALIDATION_FAILED";
    public static final String FILE_PROCESSING_ERROR_CODE = "FILE_PROCESSING_ERROR";
    public static final String FILE_PARSING_ERROR_CODE = "FILE_PARSING_ERROR";
    public static final String DATA_INTEGRITY_VIOLATION_CODE = "DATA_INTEGRITY_VIOLATION";
    public static final String OPTIMISTIC_LOCK_CONFLICT_CODE = "OPTIMISTIC_LOCK_CONFLICT";
    public static final String TRANSACTION_ERROR_CODE = "TRANSACTION_ERROR";
    public static final String INTERNAL_ERROR_CODE = "INTERNAL_ERROR";
    public static final String UNEXPECTED_ERROR_CODE = "UNEXPECTED_ERROR";
    public static final String UNHANDLED_EXCEPTION_CODE = "UNHANDLED_EXCEPTION";
    
    // ================== Log Messages ================== //
    
    /**
     * Log message templates
     */
    public static final String RESOURCE_NOT_FOUND_LOG = "Resource not found [{}]: {} - Path: {}";
    public static final String BAD_REQUEST_LOG = "Bad request [{}]: {} - Errors: {}";
    public static final String INVALID_ARGUMENT_LOG = "Invalid argument [{}]: {}";
    public static final String TYPE_MISMATCH_LOG = "Type mismatch [{}]: {}";
    public static final String CONSTRAINT_VIOLATIONS_LOG = "Constraint violations [{}]: {}";
    public static final String METHOD_ARGUMENT_VALIDATION_LOG = "Method argument validation failed [{}]: {} errors";
    public static final String FILE_PROCESSING_ERROR_LOG = "File processing error [{}]: {} - File: {}";
    public static final String FILE_PARSING_ERROR_LOG = "File parsing error [{}]: {}";
    public static final String DATA_INTEGRITY_VIOLATION_LOG = "Data integrity violation [{}]: {} - Root cause: {}";
    public static final String OPTIMISTIC_LOCKING_LOG = "Optimistic locking conflict [{}]: {} - Entity: {}";
    public static final String TRANSACTION_ERROR_LOG = "Transaction system error [{}]: {} - Root cause: {}";
    public static final String PROGRAMMING_ERROR_LOG = "CRITICAL: Programming error detected [{}] - THIS IS A BUG! Exception: {} at {}";
    public static final String FRAMEWORK_EXCEPTION_LOG = "Framework runtime exception [{}]: {} - Type: {}";
    public static final String UNEXPECTED_RUNTIME_LOG = "Unexpected runtime exception [{}]: {} - Type: {}";
    public static final String UNHANDLED_CHECKED_LOG = "Unhandled checked exception [{}] - Missing specific handler for: {} - Message: {}";
    
    /**
     * Type mismatch message template
     */
    public static final String TYPE_MISMATCH_MESSAGE_TEMPLATE = "Parameter '%s' must be of type %s";
    
    /**
     * Debug info message template
     */
    public static final String DEBUG_INFO_TEMPLATE = "Check logs with errorId: %s";
    
    // ================== Content Types ================== //
    
    public static final String DUPLICATE_KEYWORD = "duplicate";
    public static final String FOREIGN_KEY_KEYWORD = "foreign key";
    public static final String ENTITY_KEYWORD = "entity";
    
    // ================== URL Templates ================== //
    
    public static final String URI_PREFIX = "uri=";
    public static final String VALIDATION_FAILED_URI_SUFFIX = "validation-failed";
}

File: src/main/kotlin/org/myfintech/payment/exception/handler/ExceptionHandlerConstants.kt
package org.myfintech.payment.exception.handler

/**
 * Constants used in exception handling for consistent error responses.
 * Centralizes all string literals to avoid duplication and improve maintainability.
 */
object ExceptionHandlerConstants {
    // ================== Property Keys ================== //
    /**
     * Standard error response property keys
     */
    const val ERROR_ID_KEY: String = "errorId"
    const val TIMESTAMP_KEY: String = "timestamp"
    const val PATH_KEY: String = "path"
    const val ERROR_CODE_KEY: String = "errorCode"
    const val METHOD_KEY: String = "method"
    const val DEBUG_INFO_KEY: String = "debugInfo"

    /**
     * Validation-specific property keys
     */
    const val VALIDATION_ERRORS_KEY: String = "validationErrors"
    const val FIELD_ERRORS_KEY: String = "fieldErrors"
    const val GLOBAL_ERRORS_KEY: String = "globalErrors"
    const val FIELD_KEY: String = "field"
    const val MESSAGE_KEY: String = "message"
    const val REJECTED_VALUE_KEY: String = "rejectedValue"
    const val INVALID_VALUE_KEY: String = "invalidValue"
    const val OBJECT_KEY: String = "object"

    /**
     * Type mismatch property keys
     */
    const val PARAMETER_KEY: String = "parameter"
    const val EXPECTED_TYPE_KEY: String = "expectedType"
    const val ACTUAL_VALUE_KEY: String = "actualValue"

    /**
     * File processing property keys
     */
    const val FILE_NAME_KEY: String = "fileName"

    /**
     * Suggestion property key
     */
    const val SUGGESTION_KEY: String = "suggestion"

    // ================== Error Titles ================== //
    const val RESOURCE_NOT_FOUND_TITLE: String = "Resource Not Found"
    const val BAD_REQUEST_TITLE: String = "Bad Request"
    const val INVALID_ARGUMENT_TITLE: String = "Invalid Argument"
    const val TYPE_MISMATCH_TITLE: String = "Type Mismatch"
    const val VALIDATION_FAILED_TITLE: String = "Validation Failed"
    const val FILE_PROCESSING_ERROR_TITLE: String = "File Processing Error"
    const val FILE_PARSING_ERROR_TITLE: String = "File Parsing Error"
    const val DATA_CONSTRAINT_VIOLATION_TITLE: String = "Data Constraint Violation"
    const val CONCURRENT_UPDATE_CONFLICT_TITLE: String = "Concurrent Update Conflict"
    const val TRANSACTION_FAILED_TITLE: String = "Transaction Failed"
    const val INTERNAL_SERVER_ERROR_TITLE: String = "Internal Server Error"

    // ================== Error Messages ================== //
    /**
     * Generic error messages
     */
    const val UNEXPECTED_ERROR_MESSAGE: String = "An unexpected error occurred"
    const val UNEXPECTED_ERROR_WITH_NOTIFICATION: String = "An unexpected error occurred. Our team has been notified."
    const val INVALID_VALUE_DEFAULT: String = "Invalid value"
    const val INVALID_OBJECT_DEFAULT: String = "Invalid object"
    const val UNKNOWN_TYPE: String = "unknown"
    const val UNKNOWN_ENTITY: String = "Unknown entity"
    const val UNKNOWN_LOCATION: String = "unknown location"
    const val UNKNOWN_FILE: String = "unknown"
    const val NULL_VALUE: String = "null"

    /**
     * Validation error messages
     */
    const val VALIDATION_CONSTRAINTS_VIOLATED: String = "One or more validation constraints were violated"
    const val REQUEST_VALIDATION_FAILED: String = "Request validation failed. Please check the errors and try again."
    const val CLIENT_NOT_FOUND: String = "client not found: "
    const val CONTRACT_NOT_FOUND: String = "contract not found: "

    /**
     * Data integrity messages
     */
    const val DATA_CONSTRAINTS_ERROR: String = "Unable to process request due to data constraints"
    const val DUPLICATE_RECORD_ERROR: String = "A record with the same information already exists"
    const val FOREIGN_KEY_ERROR: String = "Referenced data not found or cannot be deleted due to existing dependencies"

    /**
     * Concurrency messages
     */
    const val OPTIMISTIC_LOCK_MESSAGE: String = "The record was modified by another user. Please refresh and try again."
    const val OPTIMISTIC_LOCK_SUGGESTION: String = "Refresh the data and retry your changes"

    /**
     * Transaction messages
     */
    const val TRANSACTION_COMPLETION_ERROR: String = "Unable to complete the transaction. Please try again."

    // ================== Error Codes ================== //
    const val RESOURCE_NOT_FOUND_CODE: String = "RESOURCE_NOT_FOUND"
    const val BAD_REQUEST_CODE: String = "BAD_REQUEST"
    const val INVALID_ARGUMENT_CODE: String = "INVALID_ARGUMENT"
    const val TYPE_MISMATCH_CODE: String = "TYPE_MISMATCH"
    const val VALIDATION_FAILED_CODE: String = "VALIDATION_FAILED"
    const val FILE_PROCESSING_ERROR_CODE: String = "FILE_PROCESSING_ERROR"
    const val FILE_PARSING_ERROR_CODE: String = "FILE_PARSING_ERROR"
    const val DATA_INTEGRITY_VIOLATION_CODE: String = "DATA_INTEGRITY_VIOLATION"
    const val OPTIMISTIC_LOCK_CONFLICT_CODE: String = "OPTIMISTIC_LOCK_CONFLICT"
    const val TRANSACTION_ERROR_CODE: String = "TRANSACTION_ERROR"
    const val INTERNAL_ERROR_CODE: String = "INTERNAL_ERROR"
    const val UNEXPECTED_ERROR_CODE: String = "UNEXPECTED_ERROR"
    const val UNHANDLED_EXCEPTION_CODE: String = "UNHANDLED_EXCEPTION"

    // ================== Log Messages ================== //
    /**
     * Log message templates
     */
    const val RESOURCE_NOT_FOUND_LOG: String = "Resource not found [{}]: {} - Path: {}"
    const val BAD_REQUEST_LOG: String = "Bad request [{}]: {} - Errors: {}"
    const val INVALID_ARGUMENT_LOG: String = "Invalid argument [{}]: {}"
    const val TYPE_MISMATCH_LOG: String = "Type mismatch [{}]: {}"
    const val CONSTRAINT_VIOLATIONS_LOG: String = "Constraint violations [{}]: {}"
    const val METHOD_ARGUMENT_VALIDATION_LOG: String = "Method argument validation failed [{}]: {} errors"
    const val FILE_PROCESSING_ERROR_LOG: String = "File processing error [{}]: {} - File: {}"
    const val FILE_PARSING_ERROR_LOG: String = "File parsing error [{}]: {}"
    const val DATA_INTEGRITY_VIOLATION_LOG: String = "Data integrity violation [{}]: {} - Root cause: {}"
    const val OPTIMISTIC_LOCKING_LOG: String = "Optimistic locking conflict [{}]: {} - Entity: {}"
    const val TRANSACTION_ERROR_LOG: String = "Transaction system error [{}]: {} - Root cause: {}"
    const val PROGRAMMING_ERROR_LOG: String =
        "CRITICAL: Programming error detected [{}] - THIS IS A BUG! Exception: {} at {}"
    const val FRAMEWORK_EXCEPTION_LOG: String = "Framework runtime exception [{}]: {} - Type: {}"
    const val UNEXPECTED_RUNTIME_LOG: String = "Unexpected runtime exception [{}]: {} - Type: {}"
    const val UNHANDLED_CHECKED_LOG: String =
        "Unhandled checked exception [{}] - Missing specific handler for: {} - Message: {}"

    /**
     * Type mismatch message template
     */
    const val TYPE_MISMATCH_MESSAGE_TEMPLATE: String = "Parameter '%s' must be of type %s"

    /**
     * Debug info message template
     */
    const val DEBUG_INFO_TEMPLATE: String = "Check logs with errorId: %s"

    // ================== Content Types ================== //
    const val DUPLICATE_KEYWORD: String = "duplicate"
    const val FOREIGN_KEY_KEYWORD: String = "foreign key"
    const val ENTITY_KEYWORD: String = "entity"

    // ================== URL Templates ================== //
    const val URI_PREFIX: String = "uri="
    const val VALIDATION_FAILED_URI_SUFFIX: String = "validation-failed"
}